swerve_drive_controller/src/swerve_drive_controller.cpp

```diff
@@ -0,0 +1,539 @@
+
+  std::vector<std::tuple<WheelCommand &, double, std::string>> wheel_data = {
+    {wheel_command[0], params_.front_left_velocity_threshold / params_.wheel_radius,
+     "front_left_wheel"},
+    {wheel_command[1], params_.front_right_velocity_threshold / params_.wheel_radius,
+     "front_right_wheel"},
+    {wheel_command[2], params_.rear_left_velocity_threshold / params_.wheel_radius,
+     "rear_left_wheel"},
+    {wheel_command[3], params_.rear_right_velocity_threshold / params_.wheel_radius,
+     "rear_right_wheel"}};
+
+  for (const auto & [wheel_command_, threshold, label] : wheel_data)
+  {
+    if (wheel_command_.drive_velocity > threshold)
+    {
+      wheel_command_.drive_velocity = threshold;
+    }
+  }
+
+  for (std::size_t i = 0; i < 4; i++)
+  {
+    if (!axle_handles_[i].has_value() || !wheel_handles_[i].has_value())
+    {
+      throw std::runtime_error(
+        "Axle or Wheel handle is nullptr for: " + axle_joint_names[i] + " / " +
+        wheel_joint_names[i]);
+    }
+
+    const bool is_stop = (std::fabs(linear_x_cmd) < EPS) && (std::fabs(linear_y_cmd) < EPS) &&
+                         (std::fabs(angular_cmd) < EPS);
+
+    if (is_stop)
+    {
+      axle_handles_[i]->set_position(previous_steering_angles_[i]);
+    }
+    else
+    {
+      axle_handles_[i]->set_position(wheel_command[i].steering_angle);
+      previous_steering_angles_[i] = wheel_command[i].steering_angle;
+    }
+    wheel_handles_[i]->set_velocity(wheel_command[i].drive_angular_velocity);
+  }
+
+  const auto update_dt = current_time - previous_update_timestamp_;
+  previous_update_timestamp_ = current_time;
+
+  swerve_drive_controller::OdometryState odometry_;
+
+  std::array<double, 4> velocity_array{};
+  std::array<double, 4> steering_angle_array{};
+
+  for (std::size_t i = 0; i < 4; ++i)
+  {
+    if (params_.open_loop)
+    {
+      velocity_array[i] = wheel_command[i].drive_velocity;
+      steering_angle_array[i] = wheel_command[i].steering_angle;
+    }
+    else
+    {
+      velocity_array[i] = wheel_handles_[i]->get_feedback() * params_.wheel_radius;
+      steering_angle_array[i] = axle_handles_[i]->get_feedback();
+    }
+  }
+  odometry_ = swerveDriveKinematics_.update_odometry(
+    velocity_array, steering_angle_array, update_dt.seconds());
+
+  tf2::Quaternion orientation;
+  orientation.setRPY(0.0, 0.0, odometry_.theta);
+
+  if (realtime_odometry_publisher_)
+  {
+    auto & odometry_message = realtime_odometry_publisher_->msg_;
+    odometry_message.header.stamp = time;
+    odometry_message.pose.pose.position.x = odometry_.x;
+    odometry_message.pose.pose.position.y = odometry_.y;
+    odometry_message.pose.pose.orientation.x = orientation.x();
+    odometry_message.pose.pose.orientation.y = orientation.y();
+    odometry_message.pose.pose.orientation.z = orientation.z();
+    odometry_message.pose.pose.orientation.w = orientation.w();
+    realtime_odometry_publisher_->tryPublish(odometry_message);
+  }
```

Comment:
The odometry message published here does not contain twist information (twist.linear.x, twist.linear.y, twist.angular.z). I have confirmed this issue in the Gazebo test.